Make GLFWInput and GLFWInputT into newtypes.

bch29 opened the pull request:
This change makes it a lot easier to use a GLFWInputT as part of a transformer stack that includes another StateT. I've also added a lot of derived instances (using GeneralizedNewtypeDeriving) so it's convenient to use at the top of the stack.

Commit:
Made GLFWInput and GLFWInputT into newtypes so that they can be used alongside StateT in transformer stacks. Derived many instances for the newtypes.

Commit:
Change .cabal file back to not requiring netwire-input ==0.0.4

bch29 commented:
Note that the example is currently broken, but I'll fix it dependent on what you think about my changes to netwire-input.

Not sure why it couldn't find Control.Monad.Except in the Travis build; mtl is a dependency in the .cabal file.

Mokosha commented:
Hm, I'm not totally sure what the use-cases for the changes are here...? Isn't it more transparent to use `StateT` ...? This will give you all of the standard implementations that come with the State monad without needing special implementations of everything else? I can see the argument for multiple StateT transformers, but ideally you shouldn't be directly manipulating the `GLFWInputState` regardless. I guess I'm not sure exactly what problem you're fixing? You're always free to implement your own Monad stack as a newtype and derive `MonadState MyAwesomeState`.

Mokosha commented:
Hm, looking over it I think it could work well in its own `newtype.` Make sure to rebase from master -- I've added support for GHC 7.10.

Commit:
Remove MonadRandom dependency, change GLFWInput to be GLFWInputT Identity

Commit:
Fixed example

Commit:
Revert changes to .cabal file

bch29 commented:
As you say yourself in the documentation, the GLFWInputT/GLFWInputState really shouldn't be manipulated at all other than to plumb it through in the main loop, so there's no point in exposing all of MonadState's get/set/modify/etc, especially when it comes at the cost of annoying boilerplate for the user in order to have another StateT on the stack.

In light of what you said on the netwire-input pull request, I've removed the MonadRandom derivation that I added so that MonadRandom is no longer a dependency. I've also changed GLFWInput to GLFWInputT Identity as you suggested, because it makes perfect sense.

Mokosha commented:
This looks good. Please make sure you have the latest changes from master and bump the version accordingly.

Commit:
Version bump to reflect update from master

bch29 commented:
OK, it was already rebased to upstream but I just bumped up the version number.

Mokosha commented:
Sorry, I hadn't pushed my local changes, that was my bad. If you fix the conflicts I think we're good to go.

Commit:
Merge with upstream

Commit:
Remove `Functor` dependency from `MonadMouse/Keyboard` instances by replacing `<$>` with `liftM`

Mokosha commented on netwire-input-glfw.cabal:
```diff
@@ -35,10 +35,10 @@ library
   exposed-modules:     FRP.Netwire.Input.GLFW
   -- other-modules:       
-  build-depends:       base >= 4.6 && < 4.10,
+  build-depends:       base >=4.6 && <4.9,
```
Here, too -- make sure that we don't regress the dependency version.

Mokosha commented:
Thanks for helping out!

bch29 commented:
No problem!
